lib/segment/src/payload_storage/query_checker.rs

```diff
@@ -67,55 +69,121 @@ where
     }
 }
 
-pub fn check_payload<'a, F>(
-    get_payload: F,
-    id_tracker: &IdTrackerSS,
+pub fn select_nested_indexes<'a, R>(
+    nested_path: &str,
+    field_indexes: &'a HashMap<PayloadKeyType, R>,
+) -> HashMap<PayloadKeyType, &'a Vec<FieldIndex>>
+where
+    R: AsRef<Vec<FieldIndex>>,
+{
+    let nested_prefix = format!("{}.", nested_path);
+    let nester_indexes: HashMap<_, _> = field_indexes
+        .iter()
+        .filter_map(|(key, indexes)| {
+            key.strip_prefix(&nested_prefix)
+                .map(|key| (key.into(), indexes.as_ref()))
+        })
+        .collect();
+    nester_indexes
+}
+
+pub fn check_payload<'a, R>(
+    get_payload: Box<dyn Fn() -> OwnedPayloadRef<'a> + 'a>,
+    id_tracker: Option<&IdTrackerSS>,
     query: &Filter,
     point_id: PointOffsetType,
+    field_indexes: &HashMap<PayloadKeyType, R>,
 ) -> bool
 where
-    F: Fn() -> OwnedPayloadRef<'a>,
+    R: AsRef<Vec<FieldIndex>>,
 {
     let checker = |condition: &Condition| match condition {
         Condition::Field(field_condition) => {
-            check_field_condition(field_condition, get_payload().deref())
+            check_field_condition(field_condition, get_payload().deref(), field_indexes)
         }
         Condition::IsEmpty(is_empty) => check_is_empty_condition(is_empty, get_payload().deref()),
         Condition::IsNull(is_null) => check_is_null_condition(is_null, get_payload().deref()),
-        Condition::HasId(has_id) => {
-            let external_id = match id_tracker.external_id(point_id) {
-                None => return false,
-                Some(id) => id,
-            };
-            has_id.has_id.contains(&external_id)
-        }
+        Condition::HasId(has_id) => id_tracker
+            .and_then(|id_tracker| id_tracker.external_id(point_id))
+            .map_or(false, |id| has_id.has_id.contains(&id)),
         Condition::Nested(nested) => {
-            let nested_filter = nested.filter();
-            let nested_path = JsonPathPayload::new(nested.array_key());
-            check_nested_filter(&nested_path, nested_filter, &get_payload)
+            let nested_path = nested.array_key();
+            let nested_indexes = select_nested_indexes(&nested_path, field_indexes);
+            get_payload()
+                .get_value(&nested_path)
+                .values()
+                .iter()
+                .filter_map(|value| match value {
+                    Value::Object(object) => Some(object),
+                    _ => None,
+                })
```

Comment:
```suggestion
                .filter_map(|value| value.as_object())
```

Reply:
this would effectively duplicate code in https://github.com/qdrant/qdrant/pull/1935/files#r1200739817
we should consider extracting this